tools/preflight/preflight.sh

```diff
@@ -213,70 +216,40 @@ check_storage_snapshot_class() {
-    exit_status=1
-  fi
-  # shellcheck disable=SC2143
-  if [[ $(kubectl get apiservices | grep "v1beta1.snapshot.storage.k8s.io") ]]; then
-    echo -e "${GREEN} ${CHECK} Snapshot api is in beta. No need to have default class annotation on volume snapshot class${NC}\n"
-  else
-    # shellcheck disable=SC2143
-    if [[ $(kubectl get volumesnapshotclass -o yaml | grep "is-default-class: \"true\"") ]]; then
-      echo -e "${GREEN} ${CHECK} Snapshot api is in alpha. Found a snapshot class marked as default${NC}\n"
-    else
-      echo -e "${RED} ${CROSS} Snapshot api is in alpha. No snapshot class is marked default${NC}\n"
-      exit_status=1
-    fi
-  fi
-  return ${exit_status}
-}
-
-check_feature_gates() {
-  local exit_status=0
-  local k8s_version
-  local features=()
-
-  # specially handle this for GKE alpha clusters
-  gke_all_alpha_feature="AllAlpha=true"
-
-  k8s_version=$(kubectl version --short | grep Server | awk '{print $3}')
-  echo -e "${LIGHT_BLUE}Checking if needed features are available in k8s cluster for version ${k8s_version}...${NC}\n"
-
-  k8s_version=$(echo "${k8s_version}" | cut -d '.' -f2)
-  if [[ "${k8s_version}" == "13" ]]; then
-    features=("CSIBlockVolume" "CSIDriverRegistry" "CSINodeInfo" "VolumeSnapshotDataSource")
-  else
-    features=("VolumeSnapshotDataSource")
-  fi
 
-  if [[ "${k8s_version}" -lt "15" ]]; then
-    features+=("CustomResourceWebhookConversion")
+  if [[ -z "${SNAPSHOT_CLASS}" ]]; then
+    # shellcheck disable=SC1083
+    driver=$(kubectl get sc "${STORAGE_CLASS}" | grep -E "(^|\s)${STORAGE_CLASS}($|\s)" | awk {'print $2'})
+    # shellcheck disable=SC1083
+    vsList=$(kubectl get volumesnapshotclass | grep "$driver" | awk {'print $1'})
+
+    if [[ -n "${vsList}" ]]; then
+      # shellcheck disable=SC2162
+      while read -r vs; do
+        # shellcheck disable=SC2143
+        if [[ $(kubectl get volumesnapshotclass "$vs" -o yaml | grep "is-default-class: \"true\"") ]]; then
+          SNAPSHOT_CLASS=$vs
+          break
+        fi
+        SNAPSHOT_CLASS=$vs
+      done <<<"$vsList"
+    fi
   fi
 
-  if [[ ${k8s_version} -ge "17" ]]; then
-    echo -e "${GREEN} ${CHECK} No feature gates needed${NC}\n"
+  if [[ -z "${SNAPSHOT_CLASS}" ]]; then
+    echo -e "${RED} ${CROSS} Volume snapshot class not found with same driver of given storage class${NC}\n"
+    exit_status=1
     return ${exit_status}
   fi
 
-  if [[ $(check_if_ocp) == "Y" ]]; then
-    features_enabled=$(kubectl get cm -n openshift-kube-apiserver -oyaml | grep feature)
+  # shellcheck disable=SC2143
+  if [[ $(kubectl get volumesnapshotclass | grep -E "(^|\s)${SNAPSHOT_CLASS}($|\s)") ]]; then
+    echo -e "${GREEN} ${CHECK} Volume snapshot class \"${SNAPSHOT_CLASS}\" found in cluster${NC}\n"
   else
-    features_enabled=$(kubectl get po -n kube-system -oyaml | grep feature)
+    echo -e "${RED} ${CROSS} Volume snapshot class \"${SNAPSHOT_CLASS}\" not found in cluster${NC}\n"
+    exit_status=1
```

Comment:
Do we have a check if the user given SNAPSHOT_CLASS belongs to the STORAGE_CLASS?

Reply:
no, will make changes for it.

Reply:
made changes.. please review